### PR TITLE
feat(ci): Run FeG radius unit tests in CI

### LINKED ISSUE #14350: Enable unit tests for radius in the CI
## Summary

The task #14291 and tests in #14347 showed that the radius tests are not executed in the CI. We should add CI jobs for them (or extend existing jobs).

## Acceptance criteria

* The radius unit tests in `feg/radius/src` and `feg/radius/lib` are run in the CI (the lib tests might already be in the CI). Think about running linters too.
* The new CI jobs are green.

## Tasks

- [x] Enable the tests in the CI. You can take a look at `make -C ${MAGMA_ROOT}/feg/gateway precommit` how to run Go tests.
- [x] Enable linting for the code if there is no linter yet.
- [x] Make sure everything is green. If tests are red try to fix them, but don't invest unlimited effort. You can ignore/skip tests if necessary and create a follow-up task.
- [x] Prepare internal review meeting.

### PR DESCRIPTION
**Pull request by voisey**
## Summary

Works towards https://github.com/magma/magma/issues/14350.

Runs the unit tests in `feg/radius/src` and `feg/radius/lib` in the CI. Some tests fail or are flaky so I have temporarily skipped these - I will open an issue to fix these skipped tests. I have also modified the way the tests are run by modifying the `run.sh` file, since the previous way attempts to run the tests in some paths multiple times.

This will also be followed up by a PR to run the linting in the CI.

## Test Plan

- [x] The tests are green locally
- [x] FeG Lint & Test is green in [CI](https://github.com/magma/magma/actions/runs/3647777594)

### CODE REVIEW

File: .github/workflows/feg-workflow.yml [hunk 1]
```diff
@@ -99,6 +102,18 @@ jobs:
         with:
           name: Unit Test Results
           path: "/tmp/test-results"
+      - name: Run radius src unit tests
+        if: always()
+        id: radius-src-unit-tests
+        run: |
+          cd $MAGMA_ROOT/feg/radius/src
```
Comment (LKreutzer) [inline .github/workflows/feg-workflow.yml]: Nit: Could use `working-directory` for this.

File: .github/workflows/feg-workflow.yml [hunk 2]
```diff
@@ -99,6 +102,18 @@ jobs:
...
+          ./run.sh test
+      - name: Run radius lib unit tests
+        if: always()
+        id: radius-lib-unit-tests
+        run: |
+          cd $MAGMA_ROOT/feg/radius/lib/go/
```
Comment (LKreutzer) [inline .github/workflows/feg-workflow.yml]: Same as above.

File: .github/workflows/feg-workflow.yml [hunk 3]
```diff
@@ -116,7 +131,7 @@ jobs:
       # Notify ci channel when failing
       # Plugin info: https://github.com/marketplace/actions/slack-notify
       - name: Notify failure to slack
-        if: ( steps.feg-precommit.outcome=='failure' || steps.feg-precommit-upload.outcome=='failure' ) && github.event_name == 'push'
+        if: ( steps.feg-precommit.outcome=='failure' || steps.feg-precommit-upload.outcome=='failure' || steps.radius-src-unit-tests.outcome=='failure' || steps.radius-lib-unit-tests.outcome=='failure') && github.event_name == 'push'
```
Comment (LKreutzer) [inline .github/workflows/feg-workflow.yml]: Can we split this line up please?

Comment (LKreutzer) [reply .github/workflows/feg-workflow.yml]: Use:
```
if: |
  condition \
 ...

File: feg/radius/lib/go/oc/exporters/ods_test.go [hunk 4]
```diff
@@ -83,6 +83,7 @@ func (m *mockConfigProvider) setConfig(cfg Config) Config {
 }
 
 func TestPostODS(t *testing.T) {
+	t.Skip()
```
Comment (LKreutzer) [inline feg/radius/lib/go/oc/exporters/ods_test.go]: Why is this needed?

File: feg/radius/src/modules/eap/authstate/manager.memory_test.go [hunk 5]
```diff
@@ -43,7 +43,7 @@ func performSignleReadWriteDeleteReadTest(t *testing.T, manager Manager, authReq
 	// Arrange (randomize state)
 	correlationID := rand.Intn(9999999)
 	eapType := packet.EAPTypeAKA
-	protocolState := string(rand.Intn(999999))
+	protocolState := fmt.Sprint(rand.Intn(999999))
```
Comment (LKreutzer) [inline feg/radius/src/modules/eap/authstate/manager.memory_test.go]: Sorry, I am not a go expert - what is happening here?

**Review by LKreutzer — CHANGES_REQUESTED** [review]

File: feg/radius/lib/go/oc/exporters/ods_test.go (hunk 4 above)
Comment (LKreutzer) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: Just read the description: Maybe add a comment why this is here?

File: .github/workflows/feg-workflow.yml (hunk 1 above)
Comment (voisey) [reply .github/workflows/feg-workflow.yml]: Done. Nice suggestion, thanks :slightly_smiling_face:

File: .github/workflows/feg-workflow.yml (hunk 2 above)
Comment (voisey) [reply .github/workflows/feg-workflow.yml]: Done

File: .github/workflows/feg-workflow.yml (hunk 3 above)
Comment (voisey) [reply .github/workflows/feg-workflow.yml]: Done. Looking at the other places this syntax is used, it seems the ` \`s aren't needed, but please say if I'm wrong

Comment (LKreutzer) [reply .github/workflows/feg-workflow.yml]: You are right actually, sorry for the confusion. They are only needed in Bash.

File: feg/radius/lib/go/oc/exporters/ods_test.go (hunk 4 above)
Comment (voisey) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: I've just created an issue and linked it in all cases. Does that look alright?

File: feg/radius/src/modules/eap/authstate/manager.memory_test.go (hunk 5 above)
Comment (voisey) [reply feg/radius/src/modules/eap/authstate/manager.memory_test.go]: The original code converts the random integer into a rune, which is a single unicode character, while the current state converts it into a string of integers. Note that I have just changed the function I've used to convert it to a string of integers, but the effect is the same.

Without this change, I get the following error when I run the tests:
```
modules/eap/authstate/manager.memory_test.go:46:19: conversion from int to string yields a string of one rune, not a string of digits (did you mean fmt.Sprint(x)?)
```
and the two tests where this function is used fail:
```
WARN[0000] FAIL fbc/cwf/radius/modules/eap/authstate [build failed] 
bad output from test2json: FAIL fbc/cwf/radius/modules/eap/authstate [build failed]
```

While with the change, the two tests pass. It's hard to know for certain what the intention of the original author was, but for the tests to pass any string will do. I.e. if I change it to `protocolState := ""`, the tests works.

File: feg/radius/lib/go/oc/exporters/ods_test.go (hunk 4 above)
Comment (LKreutzer) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: Could we maybe have the comment being a bit more explicit? Perhaps something like:
```
// This test is being skipped due to .... 
//  TODO https://github.com/magma/magma/issues/14659
```

Comment (LKreutzer) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: If you run these tests do they show up as skipped or successful?

Comment (voisey) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: With gotestsum you can see that they're skipped, see [here](https://github.com/magma/magma/actions/runs/3649750636/jobs/6164838947#step:13:72)

Comment (LKreutzer) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: Okay, that's good. Then I think it is just good to be on the nose in the code, so that devs immediately see what is happening.

Comment (voisey) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: What did you have in mind after the "due to"? I haven't done an analysis of the reasons they're flaky/failing yet, because I wanted to get the working set in the CI first

Comment (voisey) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: > Okay, that's good. Then I think it is just good to be on the nose in the code, so that devs immediately see what is happening.

Agreed!

Comment (LKreutzer) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: In that case just write that they are very(?) flaky.

Comment (voisey) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: I've added some more info. Could you please take a look [here](https://github.com/magma/magma/pull/14655/commits/f5e6a93f9d8413b4a6d81f8469f7278c2b0f8bad) and see what you think?

Comment (LKreutzer) [reply feg/radius/lib/go/oc/exporters/ods_test.go]: Link doesn't work, but changes lgtm

**Review by LKreutzer — APPROVED** [review]
Lgtm

**Review by MoritzThomasHuebner — APPROVED** [review]
LGTM
